gitlabform/configuration/projects_and_groups.py

```diff
@@ -18,61 +18,89 @@ def get_projects(self) -> list:
-        except KeyNotFoundException:
-            project_config = {}
-
-        logging.debug("Project config: %s" % project_config)
-
-        group, _ = group_and_project.rsplit('/', 1)
-        try:
-            group_config = self.get_config_for_group(group)
-        except ConfigNotFoundException:
-            group_config = {}
-
-        logging.debug("Group config: %s" % group_config)
-
-        try:
-            common_config = self.get("common_settings")
-        except KeyNotFoundException:
-            common_config = {}
 
+        common_config = self.get_config_common()
         logging.debug("Common config: %s" % common_config)
 
-        # this project is not in any config - return empty config
-        if not project_config and not group_config and not common_config:
-            return {}
+        group, project = group_and_project.rsplit('/', 1)
+        if '/' in group:
+            group_config = self.get_effective_subgroup_config(group)
+        else:
+            group_config = self.get_group_config(group)
+        logging.debug("Effective group/subgroup config: %s" % group_config)
 
-        # this is simplistic, but for our config format should be enough for additive merge
-        # of project, group and common configs
-
-        # first merge common config with group configs
-        for key in group_config.keys() | common_config.keys():
+        project_config = self.get_project_config(group_and_project)
+        logging.debug("Project config: %s" % project_config)
 
-            if key in common_config and key not in group_config:
-                group_config[key] = common_config[key]
-            elif key in group_config and key not in common_config:
-                group_config[key] = group_config[key]
+        common_and_group_config = self.merge_configs(common_config, group_config)
+        logging.debug("Effective config common+group/subgroup: %s" % common_and_group_config)
+
+        effective_config = self.merge_configs(common_and_group_config, project_config)
+        logging.debug("Effective config common+group/subgroup+project: %s" % effective_config)
+
+        return effective_config
+
+    def get_effective_subgroup_config(self, subgroup):
+
+        #
+        # Goes through a subgroups hierarchy, from top to bottom
+        #
+        # "x/y/x" -> ["x", "x/y", "x/y/z"]
+        #
+        # ...and for each element after 1st generate effective config from previous effective one merged with current:
+        #
+        #              |     v       |
+        #              \---> a       |
+        #                    |       v
+        #                    \------>b = effective config to return
+        #
+        # ..where a = merged_config("x", "x/y") and b = merged_config(a, "x/y/z")
+        #
+
+        effective_config = {}
+        elements = subgroup.split('/')
+        last_element = None
+        for element in elements:
+            if last_element:
```

Comment:
I think that negating this if condition to `if not last_element:` would improve code readability because it will cover first iteration

Reply:
Thanks, corrected.